docs/howto/pipeline_testing.md

```diff
@@ -0,0 +1,183 @@
+# HOWTO: Writing pipeline tests for a package
+
+## Introduction
+
+Elastic Packages are comprised of data streams. A pipeline test exercises ingest pipelines defined for a package's data stream.
+
+## Conceptual process
+
+Conceptually, running a pipeline test involves the following steps:
+
+1. Deploy the Elasticsearch instance (part of the Elastic Stack). This step takes time so it should typically be done once as a pre-requisite to running pipeline tests on multiple data streams.
+1. Upload ingest pipelines to be tested.
+1. Use [Simulate API](https://www.elastic.co/guide/en/elasticsearch/reference/master/simulate-pipeline-api.html) to process logs/metrics with the ingest pipeline.
+1. Compare generated results with expected ones.
+
+## Limitations
+
+At the moment pipeline tests have limitations. The main ones are:
+* Results of simulation can be different from real documents. They depend on the input data which may not have all fields defined.
+* There might be integrations which transform data mostly using Beats processors instead of ingest pipelines. In such cases ingest pipelines are rather plain.
+
+## Defining a pipeline test
+
+Packages have a specific folder structure (only relevant parts shown).
+
+```
+<package root>/
+  data_stream/
+    <data stream>/
+      manifest.yml
+  manifest.yml
+```
+
+To define a pipeline test we must define configuration at two levels: the package level and each dataset's level.
+
+### Dataset-level configuration
+
+Next, we must define configuration for each data stream that we want to pipeline test.
+
+```
+<package root>/
+  data_stream/
+    <data stream>/
+      _dev/
+        test/
+          pipeline/
+            (test case definitions)
+      manifest.yml
+  manifest.yml
+```
+
+### Test case definitions
+
+There are two types of test case definitions - **raw files** and **input events**.
+
+#### Raw files
+
+The raw files simplify preparing test cases using real application `.log` files. A sample log file may look like the following one for Nginx:
+
+```
+127.0.0.1 - - [07/Dec/2016:11:04:37 +0100] "GET /test1 HTTP/1.1" 404 571 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.98 Safari/537.36"
+127.0.0.1 - - [07/Dec/2016:11:04:58 +0100] "GET / HTTP/1.1" 304 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:49.0) Gecko/20100101 Firefox/49.0"
+127.0.0.1 - - [07/Dec/2016:11:04:59 +0100] "GET / HTTP/1.1" 304 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:49.0) Gecko/20100101 Firefox/49.0"
+```
+
+The pipeline test runner transforms log lines into file input events and sends to the ingest pipeline. The transformation process can be customized using a predefined configuration file:
```

Comment:
IIRC this configuration file is optional. If so, maybe say "using an optional configuration file".

Reply:
Also it's not clear where this configuration file should go in the folder structure and what it should be named.

Reply:
I've added sample filenames and mentioned them in the file tree.